### PR TITLE
16 update component codebase to match stencil style guide

### PR DESCRIPTION
**Pull request by gd2910**
## Summary of the changes
Directory structure has been changed to align with the Stencil style guide. Tests have a separate directory in the parent component and sub-directories with semantic naming of the tests they contain.

Every component has been edited to align with Stencil style guidelines, each component main file is now structured as follows;
Variables (alphabetised)
@Element
@State (alphabetised)
@Prop (alphabetised)
@Watch (with watched props behind them)
@Event (alphabetised)
Lifecycle (folowing lfiecycle process: connection, load, update, render)
@Listen
@Method
Local private methods
Render

## Related issue
#16 

## Checklist
- [ ] I have added relevant unit and visual regression tests.
- [ ] I have manually accessibility tested any changes, if relevant.
- [ ] I have ensured any changes match the Figma component library.

### CODE REVIEW

**Review by jd239 — APPROVED** [review]

**Review by MI6-451 — DISMISSED** [review]

File: packages/web-components/src/components/ic-navigation-menu/ic-navigation-menu.tsx [hunk 1]
```diff
@@ -21,22 +21,72 @@ import { getSlot, getSlotContent, isSlotUsed } from "../../utils/helpers";
   shadow: true,
 })
 export class NavigationMenu {
+  private hasButtons: boolean = false;
+  private hasNavigation: boolean = false;
+  private navItemAboveButtons: boolean = false;
+  private navGroupFirst: boolean = false;
+  private lastTabStop: HTMLElement = null;
+  private closeButton: HTMLIcButtonElement = null;
+  private navBarEl: HTMLIcTopNavigationElement;
```
Comment (GCHQ-Developer-847) [inline packages/web-components/src/components/ic-navigation-menu/ic-navigation-menu.tsx:30]: Some of these aren't in alphabetical order

File: packages/web-components/src/components/ic-navigation-menu/ic-navigation-menu.tsx [hunk 2]
```diff
@@ -21,22 +21,72 @@ import { getSlot, getSlotContent, isSlotUsed } from "../../utils/helpers";
...
+
   @Element() el: HTMLIcNavigationMenuElement;
 
   /**
-   * The version info to display.
+   * @internal - Emitted when the menu is closed.
    */
-  @Prop() version: string = "";
+  @Event() icNavigationMenuClose: EventEmitter<void>;
```
Comment (GCHQ-Developer-847) [inline packages/web-components/src/components/ic-navigation-menu/ic-navigation-menu.tsx:37]: Does this need moving under the props?

File: packages/web-components/src/components/ic-radio-option/ic-radio-option.tsx [hunk 3]
```diff
@@ -27,96 +27,69 @@ import {
 })
 export class RadioOption {
   private radioElement: HTMLInputElement;
+  private hasAdditionalField: boolean = false;
+  private defaultRadioValue: string = "";
+  private skipFocus = false;
```
Comment (GCHQ-Developer-847) [inline packages/web-components/src/components/ic-radio-option/ic-radio-option.tsx:32]: These need rearranging a bit to be in alphabetical order

File: packages/web-components/src/components/ic-search-bar/ic-search-bar.tsx [hunk 4]
```diff
@@ -48,59 +48,64 @@ let inputIds = 0;
...
-  private retryViaKeyPress: boolean;
+  private preventSubmit: boolean = false;
+  private prevNoOption: boolean = false;
   private retryButtonClick: boolean = false;
+  private retryViaKeyPress: boolean;
+  private timeoutTimer: number;
+  private truncateValue = false;
+  private searchSubmitButton: HTMLIcButtonElement;
+
+  private inputId = `ic-search-bar-input-${inputIds++}`;
```
Comment (GCHQ-Developer-847) [inline packages/web-components/src/components/ic-search-bar/ic-search-bar.tsx:67]: Do these need moving so they're in alphabetical order with the other props

File: packages/web-components/src/components/ic-search-bar/ic-search-bar.tsx [hunk 5]
```diff
@@ -109,84 +114,96 @@ export class SearchBar {
...
-   * The amount of time, in milliseconds, to wait to trigger the `icChange` event after each keystroke.
+   * Trigger loading state when fetching options asyncronously
    */
-  @Prop() debounce?: number = 0;
+  @Prop({ mutable: true }) loading?: boolean = false;
+
+  /**
+   * Change the message displayed whilst the options are being loaded externally.
+   */
+  @Prop() loadingLabel?: string = "Loading...";
```
Comment (GCHQ-Developer-847) [inline packages/web-components/src/components/ic-search-bar/ic-search-bar.tsx:166]: I think this needs swapping with loadingErrorLabel

File: packages/web-components/src/components/ic-search-bar/ic-search-bar.tsx [hunk 6]
```diff
@@ -257,85 +265,42 @@ export class SearchBar {
...
-    if (this.filteredOptions !== loadingOption)
-      this.filteredOptions = loadingOption;
-    if (this.timeout) {
-      this.timeoutTimer = window.setTimeout(() => {
-        this.filteredOptions = [
-          { label: this.loadingErrorLabel, value: "", timedOut: true },
-        ];
-      }, this.timeout);
-    }
+  private handleMouseDown = (ev: Event) => {
```
Comment (GCHQ-Developer-847) [inline packages/web-components/src/components/ic-search-bar/ic-search-bar.tsx:302]: This private method should be a bit further down

File: packages/web-components/src/components/ic-search-bar/ic-search-bar.tsx [hunk 7]
```diff
@@ -257,85 +265,42 @@ export class SearchBar {
...
+    ev.preventDefault();
   };
 
   // The icInput event is defined here so that it appears as an event for search bar
```
Comment (GCHQ-Developer-847) [inline packages/web-components/src/components/ic-search-bar/ic-search-bar.tsx:306]: debounceAriaLiveUpdate (line 340) needs moving to be with the other local private methods

File: packages/web-components/src/components/ic-select/ic-select.tsx [hunk 8]
```diff
@@ -54,193 +54,162 @@ export class Select {
   private menu: HTMLIcMenuElement;
```
Comment (GCHQ-Developer-847) [inline packages/web-components/src/components/ic-select/ic-select.tsx:54]: Some of these aren't in alphabetical order

Same with the state variables below

File: packages/web-components/src/components/ic-step/ic-step.tsx [hunk 9]
```diff
@@ -10,65 +10,66 @@ import { IcStepVariants, IcStepStatuses, IcStepTypes } from "./ic-step.types";
...
    */
-  @Prop() lastStepNum?: number;
+  @Prop() stepTitle?: string;
 
   /**
-   * @internal The name of the next step. This is managed by ic-stepper.
+   * Additional information about the step. Use this prop to override the default step status messaging displayed when selecting a step type or step status.
    */
-  @Prop() nextStepTitle?: string;
+  @Prop() stepSubtitle?: string;
```
Comment (GCHQ-Developer-847) [inline packages/web-components/src/components/ic-step/ic-step.tsx:62]: This needs to be above stepTitle

File: packages/web-components/src/components/ic-text-field/ic-text-field.tsx [hunk 10]
```diff
@@ -140,122 +160,103 @@ export class TextField {
...
-  @Prop() validationInlineInternal: boolean = false;
+  @Prop({ reflect: true }) readonly: boolean = false;
 
   /**
-   * The automatic capitalisation of the text value as it is entered/edited by the user.
-   * Available options: "off", "none", "on", "sentences", "words", "characters".
+   * If `true`, the multiline text area will be resizeable.
    */
-  @Prop() autocapitalize = "off";
+  @Prop() resize: boolean = false;
```
Comment (GCHQ-Developer-847) [inline packages/web-components/src/components/ic-text-field/ic-text-field.tsx:185]: This should be after 'required'

File: packages/web-components/src/components/ic-text-field/ic-text-field.tsx [hunk 11]
```diff
@@ -140,122 +160,103 @@ export class TextField {
...
+   * The value of the text field.
    */
-  @Prop() hiddenInput: boolean = true;
+  @Prop({ reflect: true, mutable: true }) value: string = "";
 
-  @State() numChars: number = 0;
-  @State() maxLengthExceeded: boolean = false;
-  @State() minValueUnattained: boolean = false;
-  @State() maxValueExceeded: boolean = false;
   @State() initialValue = this.value;
```
Comment (GCHQ-Developer-847) [inline packages/web-components/src/components/ic-text-field/ic-text-field.tsx:256]: Needs moving up to before the props

File: packages/web-components/src/components/ic-text-field/ic-text-field.tsx [hunk 12]
```diff
@@ -286,95 +287,43 @@ export class TextField {
...
-  private onBlur = (ev: Event) => {
-    const value = (ev.target as HTMLInputElement).value;
-    this.icBlur.emit({ value: value });
-  };
+  @Event() icInput: EventEmitter<IcValueEventDetail>;
 
   /**
    * Emitted when input gains focus.
    */
   @Event() icFocus: EventEmitter<IcValueEventDetail>;
```
Comment (GCHQ-Developer-847) [inline packages/web-components/src/components/ic-text-field/ic-text-field.tsx:312]: Think this needs to be before icInput

File: packages/web-components/src/components/ic-toast/ic-toast.tsx [hunk 13]
```diff
@@ -38,73 +38,110 @@ const TOAST_MESSAGE_CHAR_LIMIT = 140;
   shadow: true,
 })
 export class Toast {
+  dismissTimeout: number;
+  isManual: boolean;
+  interactiveElements: ActionAreaElementTypes[] = [];
```
Comment (GCHQ-Developer-847) [inline packages/web-components/src/components/ic-toast/ic-toast.tsx:43]: Needs moving to before isManual

File: packages/web-components/src/components/ic-tooltip/ic-tooltip.tsx [hunk 14]
```diff
@@ -17,27 +17,46 @@ import { onComponentRequiredPropUndefined } from "../../utils/helpers";
   shadow: true,
 })
 export class Tooltip {
+  private arrow: HTMLDivElement;
+  private delayedHideEvents = ["mouseleave"];
+  private instantHideEvents = ["focusout"];
+  private mouseOverTool: boolean = false;
+  private popperInstance: Instance;
+  private persistTooltip = false;
```
Comment (GCHQ-Developer-847) [inline packages/web-components/src/components/ic-tooltip/ic-tooltip.tsx:25]: These two need swapping

**Review by GCHQ-Developer-847 — CHANGES_REQUESTED** [review]
Looks great! Sorry about all the comments - they're only very small changes :)

File: packages/web-components/src/components/ic-hero/ic-hero.tsx [hunk 15]
```diff
@@ -35,59 +35,71 @@ import { IcHeroContentAlignments } from "./ic-hero.types";
...
 
-  /**
-   * The heading of the hero.
-   */
-  @Prop() heading!: string;
+  @State() rightContent: boolean = false;
+  @State() scrollFactor: string = "right -100px";
+  @State() leftContentFullWidth: boolean =
+    !this.rightContent && this.secondaryHeading === undefined;
+  @State() foregroundColor: IcThemeForeground = getThemeForegroundColor();
```
[the remark this answers is not in the document]
Comment (gd2910) [reply packages/web-components/src/components/ic-hero/ic-hero.tsx:42]: Have moved these around a little but leftContentFullwidth has to be below rightContent

File: packages/web-components/src/components/ic-menu/ic-menu.tsx [hunk 16]
```diff
@@ -137,34 +147,204 @@ export class Menu {
...
+    } else if (this.open) {
+      this.popperInstance.update();
+    }
+
+    if (this.open && !!this.options.length) {
+      this.setMenuScrollbar();
+    }
+  }
+
+  componentDidUpdate(): void {
```
[the remark this answers is not in the document]
Comment (gd2910) [reply packages/web-components/src/components/ic-menu/ic-menu.tsx:273]: Not that far up but it did need moving before the render (Y)

File: packages/web-components/src/components/ic-search-bar/ic-search-bar.tsx (hunk 4 above)
Comment (gd2910) [reply packages/web-components/src/components/ic-search-bar/ic-search-bar.tsx:67]: I was having problems with e2e tests involving id's so I think I isolated and forgot to move back!

File: packages/web-components/src/components/ic-text-field/ic-text-field.tsx (hunk 11 above)
Comment (gd2910) [reply packages/web-components/src/components/ic-text-field/ic-text-field.tsx:256]: this one is dependent on the value prop and will need eslint ignore when new rules incoporated

File: packages/web-components/src/components/ic-navigation-item/ic-navigation-item.tsx [hunk 17]
```diff
@@ -36,72 +36,75 @@ import OpenInNew from "../../assets/OpenInNew.svg";
...
+  private isInitialRender: boolean = true;
+  private itemEl: HTMLElement;
+
   @Element() el: HTMLIcNavigationItemElement;
 
+  @State() deviceSize: number = DEVICE_SIZES.XL;
+  @State() focusStyle = getThemeForegroundColor();
+  @State() isSideNavMobile: boolean = false;
+  @State() isTopNavChild: boolean = false;
+  @State() inTopNavSideMenu: boolean = false;
```
Comment (GCHQ-Developer-847) [inline packages/web-components/src/components/ic-navigation-item/ic-navigation-item.tsx:49]: Sorry I meant 'inTopNavSideMenu' to below 'focusStyle'

**Review by GCHQ-Developer-847 — CHANGES_REQUESTED** [review]

**Review by GCHQ-Developer-847 — APPROVED** [review]

**Review by GCHQ-Developer-741 — APPROVED** [review]

**Pull request merged**